File: ipo.cc
// IPO
#define FOR(i, a, b) for (remove_cv<remove_reference<decltype(b)>::type>::type i = (a); i < (b); i++)
#define REP(i, n) FOR(i, 0, n)

class Solution {
public:
  int findMaximizedCapital(int k, int W, vector<int>& Profits, vector<int>& Capital) {
    vector<pair<int, int>> a;
    REP(i, Profits.size())
      a.emplace_back(Capital[i], Profits[i]);
    sort(a.begin(), a.end());
    multiset<int> b;
    for (int i = 0; k--; ) {
      for (; i < a.size() && a[i].first <= W; i++)
        b.insert(a[i].second);
      if (b.empty()) break;
      W += *b.rbegin();
      b.erase(prev(b.end()));
    }
    return W;
  }
};

File: ipo.rs
// IPO
impl Solution {
    pub fn find_maximized_capital(k: i32, mut w: i32, profits: Vec<i32>, capital: Vec<i32>) -> i32 {
        use std::collections::BinaryHeap;
        let mut h = BinaryHeap::new();
        let mut a: Vec<(i32,i32)> = (0..profits.len()).map(|i| (capital[i], profits[i])).collect();
        a.sort_unstable();
        let mut i = 0;
        for _ in 0..k {
            while i < a.len() && a[i].0 <= w {
                h.push(a[i].1);
                i += 1;
            }
            if let Some(t) = h.pop() { w += t; }
            else { break; }
        }
        w
    }
}
